### PR TITLE
[BALDE-32] Implement layer stack

### PR DESCRIPTION
**Pull request by Blinkuu**

### CODE REVIEW

File: engine/src/application.cpp [hunk 1]
```diff
@@ -20,26 +21,33 @@ namespace Bald {
...
+                m_LayerStack[i - 1]->RunEvents();
+            }
+
             m_EventManager->Flush();
+
             EventManager::ClearEventQueue();
+
+            Input::InputManager::Update();
+
+            m_Window->Update();
```
Comment (BIGbadEL) [inline engine/src/application.cpp:40]: Probably updates should be called before Running events.

File: engine/src/layer_stack.h [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    private:
+        std::vector<Layer*> m_ForAddition; /**< Temp layer stack used for holding layers so that they will be pushed AFTER main loop tick */
+        unsigned m_ForAdditionAmount = 0; /**< Current number of layers in temp layer stack */
+
+        std::vector<unsigned> m_ForRemoval; /**< Temp layer stack used for holding type ids of layers that will be popped AFTER main loop tick */
+
+    }; // END OF CLASS LayerStack
+
+    template<typename L>
+    void LayerStack::PushLayer() {
```
Comment (BIGbadEL) [inline engine/src/layer_stack.h]: Consider adding the possibility to send arguments to layer constructor.

File: engine/src/layer_stack.h [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+        static_assert(std::is_base_of<Layer, L>::value, "Layer is not the base of L");
+        m_ForAddition.emplace(m_ForAddition.begin() + m_ForAdditionAmount, new L{});
+        ++m_ForAdditionAmount;
+        EventManager::Emit<LayerPushedEvent>();
+    }
+
+    template<typename L>
+    void LayerStack::PushOverlay() {
+        static_assert(std::is_base_of<Layer, L>::value, "Layer is not the base of L");
```
Comment (BIGbadEL) [inline engine/src/layer_stack.h:235]: Same as above.

File: engine/src/layer_stack.h [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        m_ForRemoval.emplace_back(Utils::get_type_id<L>());
+        EventManager::Emit<LayerPoppedEvent>();
+    }
+
+    template<typename L>
+    void LayerStack::PushLayerImmediately() {
+        CORE_LOG_INFO("[LayerStack] Pushing layer immediately...");
+
+        static_assert(std::is_base_of<Layer, L>::value, "Layer is not the base of L");
+        auto* layer = new L{};
```
Comment (BIGbadEL) [inline engine/src/layer_stack.h]: Same as above.

File: engine/src/layer_stack.h [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+
+        CORE_LOG_INFO("[LayerStack] Pushing immediately was successful");
+    }
+
+    template<typename L>
+    void LayerStack::PushOverlayImmediately() {
+        CORE_LOG_INFO("[LayerStack] Pushing overlay immediately...");
+
+        static_assert(std::is_base_of<Layer, L>::value, "Overlay is not the base of L");
+        auto* overlay = new L{};
```
Comment (BIGbadEL) [inline engine/src/layer_stack.h]: Same as above.

**Review by BIGbadEL — CHANGES_REQUESTED** [review]

File: engine/src/application.cpp (hunk 1 above)
Comment (Blinkuu) [reply engine/src/application.cpp:40]: Calling them before messes everything up. This needs further investigation.

**Review by BIGbadEL — APPROVED** [review]

**Pull request merged**